src/Raven.Server/Storage/Schema/Updates/Documents/50000/From42017.cs

```diff
@@ -1,16 +1,493 @@
+                    {
+                        PutCounterGroups(step, batch, context);
+                    }
+
+                    done = true;
+                }
+            }
+
             return true;
         }
+
+        private void PutCounterGroups(UpdateStep step, CounterBatchUpdate batch, DocumentsOperationContext context)
+        {
+            foreach (var cg in batch.Counters)
+            {
+                PutCounters(context, step, cg.Id, cg.ChangeVector, cg.Values, cg.Collection);
+            }
+
+            batch.Clear();
+        }
+
+        private static void DeleteProcessedEntries(UpdateStep step, Dictionary<string, List<long>> toDeleteByCollection)
+        {
+            foreach (var toDeleteForCollection in toDeleteByCollection)
+            {
+                var collection = new CollectionName(toDeleteForCollection.Key);
+                var table = step.WriteTx.OpenTable(CountersSchema, collection.GetTableName(CollectionTableType.CounterGroups));
+
+                foreach (var etag in toDeleteForCollection.Value)
+                {
+                    table.DeleteByIndex(CountersSchema.FixedSizeIndexes[CollectionCountersEtagsSlice], etag);
+                }
+            }
+            
+            toDeleteByCollection.Clear();
+        }
+
+        private static IEnumerable<CounterReplicationItem> GetCounters(Table table, DocumentsOperationContext ctx, string prefix)
+        {
+            if (prefix == null)
+            {
+                foreach (var result in table.SeekByPrimaryKeyPrefix(Slices.BeforeAllKeys, Slices.Empty, skip: 0))
+                {
+                    yield return CreateReplicationBatchItem(ctx, ref result.Value.Reader);
+                }
+
+                yield break;
+            }
+
+            using (DocumentIdWorker.GetSliceFromId(ctx, prefix, out Slice documentKeyPrefix, separator: SpecialChars.RecordSeparator))
+            {
+                if (table.SeekOnePrimaryKeyPrefix(documentKeyPrefix, out var reader))
+                {
+                    // first entry with prefix
+                    yield return CreateReplicationBatchItem(ctx, ref reader);
+                }
+
+                foreach (var result in table.SeekByPrimaryKeyPrefix(Slices.BeforeAllKeys, documentKeyPrefix, skip: 0))
+                {
+                    //all entries that start after prefix
+                    yield return CreateReplicationBatchItem(ctx, ref result.Value.Reader);
+                }
+            }
+        }
+
+        private readonly ObjectPool<Dictionary<LazyStringValue, PutCountersData>> _dictionariesPool
+            = new ObjectPool<Dictionary<LazyStringValue, PutCountersData>>(() => new Dictionary<LazyStringValue, PutCountersData>(LazyStringValueComparer.Instance));
+
+        private unsafe void PutCounters(DocumentsOperationContext context, UpdateStep step, string documentId, string changeVector, BlittableJsonReaderObject sourceData, string collection)
```

Comment:
This looks very similar to the `PutCounters` of the `CounterStorage`, I would prefer to reuse _if possible_.

Reply:
same for the rest of the methods here

Reply:
ok will try to reuse code from CountersStorage where possible